Align IP selection behavior with kernel

akhilles opened the pull request:
Sorting IPs by `bytes.Compare` gets rid of creation order, which can
lead to secondary IPs being chosed for masquerading and NodePort SNAT
purposes. Instead, stably sort by ifindex to preserve creation order and
give preference to primary IPs. Also, filter out secondary IPs.

This also more closely aligns with the Linux kernel's IP selection
behavior.

Fixes: #22853

```release-note
Align selection of IP addresses used for masquerading and NodePort SNAT with Linux kernel behavior, by preferring addresses assigned to the interface earlier and filtering out secondary addresses.
```

ldelossa commented on pkg/node/address_linux.go:
```diff
@@ -42,6 +41,11 @@ func firstGlobalAddr(intf string, preferredIP net.IP, family int, preferPublic b
 	addr, err := netlink.AddrList(link, family)
+	// Ensure stable order so that the same IP is chosen every time.
+	// More info in the issue GH-7637.
+	sort.SliceStable(addr, func(i, j int) bool {
```
What does this sort accomplish? `netlink.AddrList` is filtered to a `link` already, meaning all the `LinkIndex` fields on each address will be the same, producing the same list that `netlink.AddrList()` returned to begin with, no?

ldelossa replied:
Scratch that, I see `netlink.AddrList` may get called with a nil link.

ldelossa commented on pkg/node/address_linux.go:
```diff
@@ -57,16 +61,19 @@ retryScope:
 				// so if it is present we prefer this one.
 				if a.IP.Equal(preferredIP) {
 					hasPreferred = true
+				} else if a.Flags&unix.IFA_F_SECONDARY > 0 {
```
If we want to filter our secondary addresses, I think we should do it all of the time. Hoisting this to the first or second statement in for loop maybe best.

akhilles replied:
In this case, I think `preferredIP` should take precedence over primary/secondary because it's being explicitly provided as an input.

I don't have a strong preference, but respecting `preferredIP` seems less surprising.

pchaigno replied:
Maybe we could refactor the code a bit to make this clearer (note the comment update as well):
```go
// If the IP is the same as the preferredIP, that
// means that maybe it is restored from node_config.h,
// so if it is present we prefer this one, even if it
// is a secondary address.
if a.IP.Equal(preferredIP) {
	hasPreferred = true
	continue
}

if a.Flags&unix.IFA_F_SECONDARY > 0 {
	// Skip secondary addresses
	continue
}
```

ldelossa replied:
Won't this be a bit problematic in the case where you may have a preferred IP on an interface, but also a secondary IP which overlaps it? You'd potentially select the secondary IP if it was created first I believe. 

Maybe these are edge cases we aren't concerned with?

akhilles replied:
I might be misunderstanding the scenario you described, but I think the preferred IP would still be selected since it has precedence over the ordering *and* the filtering.

ldelossa replied:
What I'm trying to get it is, we must explain this function like this, currently, "We filter secondary IP addresses, unless it is preferred. If two preferred IP addresses exist across interfaces, we return the one created first". 

If we always filter secondary, the outcome is more predictable. "We filter secondary IP addresses. IPs are returned in creation order across interfaces". 

Question tho, given your statement here: https://github.com/cilium/cilium/pull/22866#discussion_r1065007987 why do you propose we filter these secondary addresses? It seems like simply sorting by creation order fixes your issue?

akhilles replied:
Filtering secondary addresses isn't needed, I added it only to be consistent with the kernel's [`inet_select_addr`](https://elixir.bootlin.com/linux/v5.11.8/C/ident/inet_select_addr) function. Maybe it's better to remove it for now? I can't think of a use-case where creation order isn't sufficient for selecting a primary IP.

pchaigno replied:
I think being consistent with the kernel is good (as long as it's in a separate commit). That avoids surprises for users when they switch from kube-proxy to KPR for example.

akhilles replied:
I pulled the filtering into a separate commit and added this comment:
> All secondary IPs, except the preferredIP, are filtered out.

IMO, this is clear enough since currently, there's only one preferred IP that's provided as an input.

pchaigno commented on pkg/node/address_linux.go:
```diff
@@ -42,6 +41,11 @@ func firstGlobalAddr(intf string, preferredIP net.IP, family int, preferPublic b
+	// More info in the issue GH-7637.
+	sort.SliceStable(addr, func(i, j int) bool {
+		return addr[i].LinkIndex < addr[j].LinkIndex
+	})
```
Any reason to perform this sort before IPs are even filtered?

akhilles replied:
I was thinking it helps with readability since we're only sorting in one place, right after fetching the addrs. I can switch back to sorting `ipsPublic` and `ipsPrivate` separately. Wdty about fetching addrs + filtering out secondary IPs + sorting in that order?

pchaigno replied:
Upside of sorting at the end is that we have less stuff to sort, though I'm not sure it makes for a significant difference.

akhilles replied:
I'll switch this back to sorting at the end since it's not a strictly related change.

pchaigno commented on pkg/node/address_linux.go:
```diff
@@ -57,16 +61,19 @@ retryScope:
 					hasPreferred = true
+				} else if a.Flags&unix.IFA_F_SECONDARY > 0 {
+					// Skip secondary addresses
+					continue
```
AFAICS, there are two independent changes here: the sort and the filter for secondary addresses. In your commit you mention that this fixes an issue with floating IPs in your case. Which change is fixing your issue exactly?

Given they are independent changes, I would also prefer to have them in separate commits. We may need to backport one change without backporting the other.

akhilles replied:
The sort by creation time is enough to fix our issue, because the floating IP is assigned by an agent on the node after the primary IP has been assigned.

I'll separate them into separate commits.

pchaigno replied:
> The sort by creation time is enough to fix our issue, because the floating IP is assigned by an agent on the node after the primary IP has been assigned.

That's what I don't understand. If the floating IP is assigned to the same interface as the primary address, why does sorting by interface indexes change anything? 

> I'll separate them into separate commits.

Thanks!

akhilles replied:
Sorting by ifindex doesn't solve the problem, but removing the sort by `bytes.Compare` does. Sorting by `bytes.Compare` gets rid of creation order, which leads to secondary IPs being chosen instead of primary IPs. The sorting was added to address https://github.com/cilium/cilium/issues/7637#issuecomment-482152467, which actually mentions sorting by ifindex as a potential fix but the PR that resolved that issue ended up sorting using `bytes.Compare` instead. Maybe because ifindex wasn't available?

So, sorting by ifindex is an alternative to sorting by byte comparison that preserves creation order of IPs for any given interface.

I should probably clarify this in the commit message.

pchaigno replied:
Ah, I see that makes sense 👍 

Yep, an explanation of that in the commit description would be great!

pchaigno reviewed (commented):
> Previously, the IPs were sorted and the first value was chosen. This could cause a floating IP to be chosen for SNAT use instead of the primary IP address.
> 
> This patch stably sorts addresses by link index, which preserves the ordering of IPs for any given interface.

Could you give a bit more detail here? IIUC, it's the sort that is fixing your issue with floating IP addresses. How come? Is it because your floating IPs are always assigned to an interface that is created after the interface with the primary address?

> ```release-note
> Prefer primary IPs for masquerade, NodePort address selection
> ```

Maybe reword to:
```
Align selection of IP address for masquerading on Linux's logic, by preferring older interfaces and filtering out secondary addresses.
```

akhilles committed:
Sort IPs by ifindex instead of byte comparison

Sorting IPs by `bytes.Compare` gets rid of creation order, which can
lead to secondary IPs being chosed for masquerading and NodePort SNAT
purposes. Instead, stably sort by ifindex to preserve creation order and
give preference to primary IPs.

This also more closely aligns with the Linux kernel's IP selection
behavior.

Signed-off-by: Akhil Velagapudi <4@4khil.com>

akhilles committed:
Filter out secondary IPs during IP selection

This further aligns IP selection with the Linux kernel, which results in
less surprising behavior when switching from kube-proxy to KPR.

Signed-off-by: Akhil Velagapudi <4@4khil.com>

pchaigno reviewed (approved):
Thanks!

ldelossa reviewed (approved):
LGTM. Thanks.
